src/yki/handler/exam_date.clj

```diff
@@ -49,146 +31,127 @@
+                    (ok {:id id}))
+                  (do (log/error "Creating exam date failed")
+                      (internal-server-error
+                        {:success false
+                         :error   "Creating exam date failed"})))))))
 
       (context "/:id" []
         (GET "/" []
           :path-params [id :- ::ys/id]
           :return ::ys/single-exam-date-response
           (if-let [exam-date (exam-date-db/get-exam-date-by-id db id)]
             (ok {:date exam-date})
-            (do (log/error "Could not get an exam date with id " id)
+            (do (log/error "Could not get exam date" id)
                 (not-found {:success false
                             :error   "Exam date not found"}))))
 
+        (PUT "/" request
+          :path-params [id :- ::ys/id]
+          :body [exam-date-update ::ys/exam-date-type]
+          :return ::ys/response
+          (if-let [exam-date (exam-date-db/get-exam-date-by-id db id)]
+            (let [{date       :exam_date
+                   reg-start  :registration_start_date
+                   reg-end    :registration_end_date
+                   post-start :post_admission_start_date
+                   post-end   :post_admission_end_date} exam-date-update
+
+                  is-period-valid (is-sorted (filter some? (vec [reg-start reg-end post-start post-end date])))
+                  existing-date   (first (exam-date-db/get-exam-dates-by-date db (c/from-long date)))
+
+                  original-languages (vec (:languages exam-date))
+                  provided-languages (:languages exam-date-update)
+                  new-languages (set/difference (set provided-languages) (set original-languages))
+                  removable-languages (set/difference (set original-languages) (set provided-languages))
+
+                  is-language-change (> (count (concat new-languages removable-languages)) 0)
+                  has-exam-sessions (> (:count (exam-date-db/get-exam-date-session-count db id)) 0)]
+              (cond
+                (not is-period-valid) (do (log/error "Invalid exam date, registration or post admission period")
+                                          (conflict {:success false
+                                                     :error   "Invalid exam date, registration or post admission period"}))
+
+                (and (not (nil? existing-date)) (not= id (:id existing-date)))
+                (do (log/error "Another exam date with date" date "already exists")
+                    (conflict {:success false
+                               :error   "Another exam date with the same date already exists"}))
+
+                (and is-language-change has-exam-sessions)
+                (do (log/error "Changing languages for an exam date with exam sessions not allowed")
+                    (conflict {:success false
+                               :error   "Changing languages for an exam date with exam sessions not allowed"}))
```

Comment:
Jäin pohtimaan onko tämä hyvä rajoite.. mielestäni hieman epäkonsistenttia että mm. tutkintopäivän ajankohdan muuttaminen on ok vaikka olisikin jo tutkintotilaisuuksia ko. päivällä, mutta kielten vaihtaminen ei ole.

Reply:
Kielitasoja nyt ei yksinkertaisesti taideta voida antaa virkailijan muokata, mikäli järjestäjät luoneet jo kielitasokohtaisesti tilaisuuksia? Itse päivääkään ei kyllä ehkä tällöin pitäisi antaa muuttaa, kun se kuitenkin varmasti vaatisi laajempaa setvimistä järjestäjien kanssa, että päivä nyt ei olekaan se mikä sen on alun perin pitänyt olla. Mutta teoriassa näkisin sen realistisempana? Tosin voisi muuttaa niinkin että myöskään päivää ei pysty muuttamaan sen jälkeen kun tutkintopäivää koskien on jo tilaisuuksia.

Reply:
Niin, siis se on mielestäni täysin ok että tutkintopäivää voi muokata kun sille ei vielä ole tilaisuuksia. Jos tässä halutaan estää tutkintopäivän kielten muokkaaminen silloin kun tilaisuuksia on olemassa, sekin on varmaan ok mutta samalla voisi sitten olla konsistentti ja evätä itse päivämäärän muokkauksen (silloin kun niitä tilaisuuksia ko. päivälle on).